aten/src/ATen/Error.h

```diff
@@ -25,262 +25,118 @@
-  // by `backtrace_symbols` is malloc'd and must be manually freed, but not the
-  // strings inside the array.
-  std::unique_ptr<char*, std::function<void(char**)>> raw_symbols(
-      ::backtrace_symbols(callstack.data(), callstack.size()),
-      /*deleter=*/free);
-  const std::vector<std::string> symbols(
-      raw_symbols.get(), raw_symbols.get() + callstack.size());
-
-  // The backtrace string goes into here.
-  std::ostringstream stream;
-
-  for (size_t frame_number = 0; frame_number < callstack.size();
-       ++frame_number) {
-    const auto frame = parse_frame_information(symbols[frame_number]);
+inline void _str(std::stringstream& /*ss*/) {}
 
-    // frame #<number>:
-    stream << "frame #" << frame_number << ": ";
-
-    if (!frame.has_value()) {
-      // In the edge-case where we couldn't parse the frame string, we can just
-      // use it directly (it may have a different format).
-      stream << symbols[frame_number] << "\n";
-    } else {
-      // <function_name> + <offset> (<return-address> in <object-file>)
-      stream << frame->function_name << " + " << frame->offset_into_function
-             << " (" << callstack[frame_number] << " in " << frame->object_file
-             << ")\n";
-    }
-  }
-
-  return stream.str();
+template <typename T>
+inline void _str(std::stringstream& ss, const T& t) {
+  ss << t;
 }
-#endif // !defined(_WIN32)
 
-/// A tiny implementation of static `all_of`.
-template <bool...>
-struct pack;
-template <bool... values>
-struct all_of : std::is_same<pack<values..., true>, pack<true, values...>> {};
+template <typename T, typename... Args>
+inline void
+_str(std::stringstream& ss, const T& t, const Args&... args) {
+  _str(ss, t);
+  _str(ss, args...);
+}
 
-/// A printf wrapper that returns an std::string.
-inline std::string format(const char* format_string, ...) {
-  static constexpr size_t kMaximumStringLength = 4096;
-  char buffer[kMaximumStringLength];
+} // namespace detail
 
-  va_list format_args;
-  va_start(format_args, format_string);
-  vsnprintf(buffer, sizeof(buffer), format_string, format_args);
-  va_end(format_args);
+// Convert a list of string-like arguments into a single string.
+template <typename... Args>
+inline std::string str(const Args&... args) {
+  std::stringstream ss;
+  detail::_str(ss, args...);
+  return ss.str();
+}
 
-  return buffer;
+// Specializations for already-a-string types.
+template <>
+inline std::string str(const std::string& str) {
+  return str;
+}
+inline std::string str(const char* c_str) {
+  return c_str;
 }
 
 /// Represents a location in source code (for debugging).
 struct SourceLocation {
-  std::string toString() const {
-    return format("%s at %s:%d", function, file, line);
-  }
-
   const char* function;
   const char* file;
   uint32_t line;
 };
-} // namespace detail
+
+inline std::ostream& operator<<(std::ostream& out, const SourceLocation& loc) {
```

Comment:
We could also just move all implementations into `Error.cpp` file since we already have it. This code is definitely not performance critical

Reply:
I'm thinking it might be handy to have `at::str` available for quick and dirty stringification. I guess everything else can go in the cpp file though.